pkg/server/handler/sso_customtoken.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	// generate access-token
+	token, err := store.NewToken(payload.AppName, info.ID)
+	if err != nil {
+		panic(err)
+	}
+
+	if err = store.Put(&token); err != nil {
+		panic(err)
+	}
+
+	authResponse, err := AuthResponseFactory{
+		AssetStore: h.AssetStore,
+		Conn:       payload.DBConn,
+	}.NewAuthResponse(info, user, token.AccessToken, payload.HasMasterKey())
+	if err != nil {
+		response.Err = skyerr.MakeError(err)
+		return
+	}
+
+	// Populate the activity time to user
+	now := timeNow()
+	info.LastSeenAt = &now
+	if err := payload.DBConn.UpdateAuth(&info); err != nil {
+		response.Err = skyerr.MakeError(err)
+		return
+	}
+
+	// update user record last login time
+	user.UpdatedAt = now
+	user.UpdaterID = info.ID
+	user.Data[UserRecordLastLoginAtKey] = now
+	if err := payload.Database.Save(&user); err != nil {
+		response.Err = skyerr.MakeError(err)
+		return
+	}
+
+	response.Result = authResponse
+}
+
+func (h *SSOCustomTokenLoginHandler) handleLogin(payload *router.Payload, p *ssoCustomTokenLoginPayload, authinfo *skydb.AuthInfo, user *skydb.Record) skyerr.Error {
+	principalID := p.Claims.Subject
+	var customTokenInfo skydb.CustomTokenInfo
+	createNewUser := false
+
+	if err := payload.DBConn.GetCustomTokenInfo(principalID, &customTokenInfo); err != nil {
+		if err != skydb.ErrUserNotFound {
+			return skyerr.MakeError(err)
+		}
+		createNewUser = true
+		*authinfo = skydb.NewAnonymousAuthInfo()
+	}
+
+	if !createNewUser {
+		if err := payload.DBConn.GetAuth(customTokenInfo.UserID, authinfo); err != nil {
+			if err != skydb.ErrUserNotFound {
+				return skyerr.MakeError(err)
+			}
+			createNewUser = true
+			*authinfo = skydb.NewAnonymousAuthInfo()
+		}
+	}
+
+	userRecordContext := &authUserRecordContext{
+		DBConn:         payload.DBConn,
+		Database:       payload.Database,
+		AssetStore:     h.AssetStore,
+		HookRegistry:   h.HookRegistry,
+		AuthRecordKeys: h.AuthRecordKeys,
+		Context:        payload.Context,
+	}
+	if createNewUser {
+		userRecordContext.BeforeSaveFunc = func(conn skydb.Conn, info *skydb.AuthInfo) error {
+			if err := conn.CreateAuth(info); err != nil {
+				if err == skydb.ErrUserDuplicated {
+					return errUserDuplicated
+				}
+
+				return skyerr.MakeError(err)
+			}
+			return nil
+		}
+		userRecordContext.BeforeCommitFunc = func(conn skydb.Conn, user *skydb.Record) error {
+			now := timeNow()
+			if err := conn.CreateCustomTokenInfo(&skydb.CustomTokenInfo{
+				UserID:      user.ID.Key,
+				PrincipalID: principalID,
+				CreatedAt:   &now,
+			}); err != nil {
+				return skyerr.MakeError(err)
+			}
+			return nil
+		}
+
+	} else {
+		userRecordContext.BeforeSaveFunc = func(conn skydb.Conn, info *skydb.AuthInfo) error {
+			if err := conn.UpdateAuth(info); err != nil {
+				return skyerr.MakeError(err)
+			}
+			return nil
+		}
+		userRecordContext.BeforeCommitFunc = func(conn skydb.Conn, user *skydb.Record) error {
+			now := timeNow()
+			if err := conn.CreateCustomTokenInfo(&skydb.CustomTokenInfo{
```

Comment:
Need further check if we need createCustomTokenInfo in this case, the other part looks great!:)